runtime/lua/vim/treesitter/query.lua

```diff
@@ -0,0 +1,241 @@
+local a = vim.api
+local lang = require'vim.treesitter.language'
+
+-- query: pattern matching on trees
+-- predicate matching is implemented in lua
+local Query = {}
+Query.__index = Query
+
+local M = {}
+
+local magic_prefixes = {['\\v']=true, ['\\m']=true, ['\\M']=true, ['\\V']=true}
+local function check_magic(str)
+  if string.len(str) < 2 or magic_prefixes[string.sub(str, 1, 2)] then
+    return str
+  end
+  return '\\v'..str
+end
+
+-- Some treesitter grammars extend others.
+-- We can use that to import the queries of the base language
+local base_language_map = {
+  cpp = {'c'},
+  typescript = {'javascript'},
+  tsx = {'typescript', 'javascript'},
+}
+
+--- Register a language as using base_language as it's base.
+--
+-- @param language The language
+-- @param base_langue The language of which the queries will be used as a base.
+function M.base_language_add(language, base_language)
+  if not base_language_map[language] then
+    base_language_map[language] = {}
+  end
+
+  table.insert(base_language_map[language], base_language)
+end
+
+--- Returns the base languages of a given language
+--
+-- @param language The language
+--
+-- @returns A table containing the languages @param language uses as it's base.
+function M.base_language_get(language)
+  return base_language_map[language] or {}
+end
+
+--- Parses a query.
+--
+-- @param language The language
+-- @param query A string containing the query (s-expr syntax)
+--
+-- @returns The query
+function M.parse_query(language, query)
+  lang.require_language(language)
+  local self = setmetatable({}, Query)
+  self.query = vim._ts_parse_query(language, vim.fn.escape(query,'\\'))
+  self.info = self.query:inspect()
+  self.captures = self.info.captures
+  self.regexes = {}
+  for id, preds in pairs(self.info.patterns) do
+    local regexes = {}
+    for i, pred in ipairs(preds) do
+      if (pred[1] == "match?" and type(pred[2]) == "number"
+          and type(pred[3]) == "string") then
+        regexes[i] = vim.regex(check_magic(pred[3]))
+      end
+    end
+    if not vim.tbl_isempty(regexes) then
+      self.regexes[id] = regexes
+    end
+  end
+  return self
+end
+
+local function read_query_files(filenames)
+  local contents = {}
+
+  for _, filename in ipairs(filenames) do
+    vim.list_extend(contents, vim.fn.readfile(filename))
+  end
+
+  return table.concat(contents, '\n')
+end
+
+--- Gets a fully fledged query from runtime files
+--
+-- @param lang The source language
+-- @param query_name The name of the query (without `.scm`)
+--
+-- @returns The query
+function M.get_query(language, query_name)
```

Comment:
perhaps query name should allow for a complete filename to be passed to it? So you can specify some random file as well? 

You can check if that file exists and then read from that one if it does.

This might solve @teto 's concern.

Reply:
Should be good now, the `query_name` will be used as is if it is a readable file.